hypershift-operator/controllers/hostedClusterInfrastructure/hostedclusterinfrastructure_controller.go

```diff
@@ -0,0 +1,286 @@
+
+	// Retrieve the HostedClusterInfrastructure custom resource
+	var hci hyperv1.HostedClusterInfrastructure
+	if err := r.Get(ctx, req.NamespacedName, &hci); err != nil {
+		log.Info("Resource was deleted")
+		return ctrl.Result{}, nil
+	}
+
+	// Check for previous work and set Phase=READY if complete
+	infraStatus := meta.IsStatusConditionTrue(hci.Status.Conditions, string(hyperv1.PlatformConfigured))
+	iamStatus := meta.IsStatusConditionTrue(hci.Status.Conditions, string(hyperv1.PlatformIAMConfigured))
+	if infraStatus && iamStatus && hci.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("Nothing to do, Hosted Cluster Infrastructure is %s", string(hyperv1.AsExpected)))
+		if hci.Status.Phase != hyperv1.PhaseReady {
+			oHci := hci.DeepCopy()
+			hci.Status.Phase = hyperv1.PhaseReady
+			if err := r.Client.Status().Patch(r.ctx, &hci, client.MergeFrom(oHci)); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check for a provider secret and exit
+	var providerSecret corev1.Secret
+	var err error
+
+	providerSecretName := hci.Spec.CloudProvider.Name
+	// Nothing to do if the cloudProvider secret name is not present
+	if providerSecretName == "" {
+		log.Info("No provider secret name, can't do a reconcile")
+		r.updateStatusConditionsOnChange(&hci, nil,
+			hyperv1.ProviderSecretConfigured,
+			metav1.ConditionFalse,
+			"No cloudProvider secret reference provided, create a cloudProvider secret in the namespace "+hci.Namespace,
+			hyperv1.NotAsExpected)
+
+		return ctrl.Result{}, nil
+	}
+
+	// Attempt to retreive the cloudProvider secret
+	err = r.Client.Get(ctx, types.NamespacedName{Namespace: hci.Namespace, Name: providerSecretName}, &providerSecret)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Could not retrieve the provider secret: %s", providerSecretName))
+		return ctrl.Result{RequeueAfter: 30 * time.Second, Requeue: true},
+			r.updateStatusConditionsOnChange(&hci, nil,
+				hyperv1.ProviderSecretConfigured,
+				metav1.ConditionFalse,
+				"The secret "+providerSecretName+" could not be retreived from namespace "+hci.Namespace,
+				hyperv1.NotAsExpected)
+	}
+	if err := r.updateStatusConditionsOnChange(&hci, nil, hyperv1.ProviderSecretConfigured, metav1.ConditionTrue, "Retreived secret "+providerSecretName, string(hyperv1.AsExpected)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Add a finalizer so the resource can't be deleted till the infrastructure is cleaned up
+	// Also set the Infra-ID if its not set.
+	if !controllerutil.ContainsFinalizer(&hci, DestroyFinalizer) {
+		controllerutil.AddFinalizer(&hci, DestroyFinalizer)
+
+		if hci.Spec.InfraID == "" {
+			hci.Spec.InfraID = hci.Name
+		}
+		if hci.Spec.SubDomain == "" {
+			hci.Spec.SubDomain = hci.Spec.InfraID
+		}
+		if err := r.patchHCI(&hci); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update infra-id: \"%s\" and error: %w", hci.Spec.InfraID, err)
+		}
```

Comment:
infraID and subDomain are immutable fields. is it possible to update them after creation?

Reply:
This is how HostedCluster_Controller treats the infra-id, so I was following this approach.
hostedCluster_types.go defines infra-id as `optional and immutable`. 
The following is the code in HostedCluster_controller that sets the infra-id if not provided [link](https://github.com/openshift/hypershift/blob/1f70f40c15c6fef0d0ccfc96ba7752c77d6873cc/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L3870)

Reply:
ah, alrighty. My primary concern was that the `immuntable` marker might prevent your logic from working the way you wanted. But if it works to set the InfraID when it's "", then I have no concerns